File: oreon-11-rp1/wgrib/testbin.c
#include <stdio.h>
#include <math.h>
#define nx	192
#define ny	94
#define dx	3
#define dy	3
/* v1.2 wesley ebisuzaki */
#define INT int
/* if the default integer is 16 bit then */
/* #define INT long int */

void main() {
	FILE *in;
	float land[nx];
	INT header;
	int i, j;
	char c;

	printf("Test creation of a binary file\n");
	printf("before running this program, enter wgrib land.grb -d 1\n");

	if ((in = fopen("dump","rb")) == NULL) {
		printf("could not find file: dump\n");
		printf("did you run wgrib land.grb -d 1?\n");
		exit(8);
	}

	if (fread((void *) &header, sizeof(INT), 1, in) != 1) {
		printf("no data?\n");
		exit(8);
	}

	if (header != nx*ny*sizeof(float)) {
		printf("wrong header size: %d\n", (int) header);
		exit(8);
	}

	for (j = 0; j < ny; j++) {
		if (fread((void *) land, sizeof(float), nx, in) != nx) {
			printf("not enough data\n");
			exit(8);
		}
		for (i = 0; i < nx; i++) {
			c = '?';
			if (floor(land[i]*10000.0+0.5) == 0) c = ' ';
			if (floor(land[i]*10000.0+0.5) == 10000.0) c = 'x';
			if (c == '?') {
				printf("bad value: %g\n",land[i]);
				exit(8);
			}
			if (i % dx == 0 && j % dy == 0) {
				putchar(c);
			}
		}
		if (j % dy == 0) putchar('\n');
	}
	printf("should see continents -- if you do, it works\n");
}

File: oreon-11-rp1/wgrib/testbin.f
c
c	sample program to read binary file in fortran
c	this may or may not work! (system dependent)
c	often works with UNIX fortrans
c
	real array(192,94)
	character*1 line(192)

	write(*,*) 'should first run wgrib -d 1 land.grb'

	open(unit=1,file='dump',form='unformatted',status='old')
	read(1) array
	err=1e-6
	do 100 j = 1, 94
	    do 90 i = 1, 192
	        if (array(i,j).eq.0.0) then
		    line(i) = ' '
		else if (abs(array(i,j)-1.0).lt.err) then
		    line(i) = 'x'
		else
		    write(*,*) 'bad values:',array(i,j)
		    stop
		endif
90	    continue
	    if (mod(j,3).eq.0) write(*,*) (line(i),i=1,192,3)
100	continue
	write(*,*) 'should see continents'
	stop
	end
